apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+websites hosted at a number of different domain names. We can call these
+"sites". Thus, the code.org product consists of the following "sites": code.org,
+studio.code.org, hourofcode.org, csedweek.org, etc.
+
+While these sites may talk to the same backend databases, point to the same
+application servers, share authentication tokens, and reuse the same code, from
+the point of view of the front-end code base, these are distinct sites.
+
+While noting what types of resources these sites might share, it's important to
+also note which resources they absolutely do not share: client side browser
+resource caches (assuming the resources are being loaded from the same domain
+name as the site), state stored across page loads in html5 local storage, single
+domain cookies, and probably some other stuff.
+
+**pages**
+
+One level below sites are pages. A single site typically consists of multiple
+pages with different url patterns. Each page might provide a distinct set of
+features to a distinct set of users.
+
+The really important aspect of a "page" is that it represents the outermost
+scope of our javascript runtime. That means any javascript being executed on
+that page is running in a shared memory environment where arbitrary values can
+be stored on the window object and be accessed by other javascript running on
+the same page. However, when the user moves to a different page, any state being
+maintained within that page's scope is wiped clean.
+
+**entry point**
+
+In theory, each page should have exactly one javascript file loaded on it (via a
+`<script>` tag) which represents the primary point of entry into our javascript
+code base. The javascript build system is responsible for aggregating all the
+necessary code into this single javascript file that is needed to setup
+everything on the page.
+
+Note 1: The build system can produce auxiliary files that must be loaded on the
+page for the "entry point" to work. These other files are not "entry points",
+but rather "common chunks" of shared code that are only separated to optimize
+page load times. These "common chunks" don't actually do anything except provide
+functionality that can be optionally used by the entry point.
+
+Note 2: We are pretty far from having a single "entry point" per page, but this
+is really what we should be striving towards.
+
+The other critical aspect of an entry point is that it should explicitly "own"
+all the application state that is needed to render and update the page. In the
+event that any of the entry point's dependencies need to access or modify the
+application state owned by the entry point, it is up to the entry point to
+explicitly pass that state to it's dependencies, along with mutators for
+modifying that state.
+
+**library**
+
+Each entry point can make use of numerous "libraries" to setup and run the
+page. A library is just an arbitary collection of related code that provides
```

Comment:
Let's assume together that nit-picks over the meanings of words (such as you have done here) are fair game, but that Paul is going to run a spelling / grammar check over this at some point after we agree to do it and before we developer-blog-publish it.